README.md
# Prometheus client for Node.js [![Build Status](https://travis-ci.org/siimon/prom-client.svg?branch=master)](https://travis-ci.org/siimon/prom-client) [![Build status](https://ci.appveyor.com/api/projects/status/k2e0gwonkcee3lp9/branch/master?svg=true)](https://ci.appveyor.com/project/siimon/prom-client/branch/master) [![Actions Status](https://github.com/siimon/prom-client/workflows/Node.js%20CI/badge.svg?branch=master)](https://github.com/siimon/prom-client/actions)

A prometheus client for Node.js that supports histogram, summaries, gauges and
counters.

## Usage

See example folder for a sample usage. The library does not bundle any web
framework. To expose the metrics, response to Prometheus HTTP requests with
the result of `registry.metrics()`.

### Usage with Node.js's `cluster` module

Node.js's `cluster` module spawns multiple processes and hands off socket
connections to those workers. Returning metrics from a worker's local registry
will only reveal that individual worker's metrics, which is generally
undesirable. To solve this, you can aggregate all of the workers' metrics in the
master process. See `example/cluster.js` for an example.

Default metrics use sensible aggregation methods. Custom metrics are summed
across workers by default. To use a different aggregation method, set the
`aggregator` property in the metric config to one of 'sum', 'first', 'min',
'max', 'average' or 'omit'. (See `lib/metrics/version.js` for an example.)

If you need to expose metrics about an individual worker, you can include a
value that is unique to the worker (such as the worker ID or process ID) in a
label. (See `example/server.js` for an example using
`worker_${cluster.worker.id}` as a label value.)

Metrics are aggregated from the global registry by default. To use a different
registry, call
`client.AggregatorRegistry.setRegistries(registryOrArrayOfRegistries)` from the
worker processes.

## API

### Default metrics

There are some default metrics recommended by Prometheus
[itself](https://prometheus.io/docs/instrumenting/writing_clientlibs/#standard-and-runtime-collectors).
To collect these, call `collectDefaultMetrics`. In addition, some
Node.js-specific metrics are included, such as event loop lag, active handles,
GC and Node.js version. See [lib/metrics](lib/metrics) for a list of all
metrics.

NOTE: Some of the metrics, concerning File Descriptors and Memory, are only
available on Linux.

`collectDefaultMetrics` optionally accepts a config object with following entries:

- `prefix` an optional prefix for metric names. Default: no prefix.
- `registry` to which metrics should be registered. Default: the global default registry.
- `gcDurationBuckets` with custom buckets for GC duration histogram. Default buckets of GC duration histogram are `[0.001, 0.01, 0.1, 1, 2, 5]` (in seconds).
- `eventLoopMonitoringPrecision` with sampling rate in milliseconds. Must be greater than zero. Default: 10.

To register metrics to another registry, pass it in as `register`:

```js
const client = require('prom-client');
const collectDefaultMetrics = client.collectDefaultMetrics;
const Registry = client.Registry;
const register = new Registry();
collectDefaultMetrics({ register });
```

To use custom buckets for GC duration histogram, pass it in as `gcDurationBuckets`:

```js
const client = require('prom-client');
const collectDefaultMetrics = client.collectDefaultMetrics;
collectDefaultMetrics({ gcDurationBuckets: [0.1, 0.2, 0.3] });
```

To prefix metric names with your own arbitrary string, pass in a `prefix`:

```js
const client = require('prom-client');
const collectDefaultMetrics = client.collectDefaultMetrics;
const prefix = 'my_application_';
collectDefaultMetrics({ prefix });
```

You can get the full list of metrics by inspecting
`client.collectDefaultMetrics.metricsList`.

### Custom Metrics

All metric types have two mandatory parameters: `name` and `help`. Refer to
<https://prometheus.io/docs/practices/naming/> for guidance on naming metrics.

For metrics based on point-in-time observations (e.g. current memory usage, as
opposed to HTTP request durations in a histogram), you should use
`registry.registerCollector(fn)`, where `fn` is a callback invoked when
Prometheus scrapes your registry. `fn` should be synchronous so that the
metric's value is current. See **Gauge** below for an example. (In particular,
you should not update metric values in a `setInterval` callback.)

See **Labels** for information on how to configure labels for all metric types.

#### Counter

Counters go up, and reset when the process restarts.

```js
const client = require('prom-client');
const counter = new client.Counter({
  name: 'metric_name',
  help: 'metric_help'
});
counter.inc(); // Increment by 1
counter.inc(10); // Increment by 10
```

#### Gauge

Gauges are similar to Counters but a Gauge's value can be decreased.

```js
const client = require('prom-client');
const gauge = new client.Gauge({ name: 'metric_name', help: 'metric_help' });
gauge.set(10); // Set to 10
gauge.inc(); // Increment 1
gauge.inc(10); // Increment 10
gauge.dec(); // Decrement by 1
gauge.dec(10); // Decrement by 10
```

There's a utility to set a gauge to the current time:

```js
gauge.setToCurrentTime();
```

And there's a utility to time durations:

```js
const end = gauge.startTimer();
xhrRequest(function(err, res) {
  end(); // Sets value to xhrRequest's duration in seconds
});
```

If the gauge is used for a point-in-time observation, you should register a
collector function:

```js
const client = require('prom-client');
const gauge = new client.Gauge({ name: 'metric_name', help: 'metric_help' });
client.register.registerCollector(() => {
  // Invoked when the registry collects its metrics' values.
  // This should be synchronous.
  gauge.set(/* the current value */);
});
```

#### Histogram

Histograms track sizes and frequency of events.

**Configuration**

The defaults buckets are intended to cover usual web/RPC requests, but they can
be overridden.

```js
const client = require('prom-client');
new client.Histogram({
  name: 'metric_name',
  help: 'metric_help',
  buckets: [0.1, 5, 15, 50, 100, 500]
});
```

Examples

```js
const client = require('prom-client');
const histogram = new client.Histogram({
  name: 'metric_name',
  help: 'metric_help'
});
histogram.observe(10); // Observe value in histogram
```

Utility to observe request durations:

```js
const end = histogram.startTimer();
xhrRequest(function(err, res) {
  const seconds = end(); // Observes and returns the value to xhrRequests duration in seconds
});
```

#### Summary

Summaries calculate percentiles of observed values.

**Configuration**

The default percentiles are: 0.01, 0.05, 0.5, 0.9, 0.95, 0.99, 0.999. But they
can be overridden by specifying a `percentiles` array:

```js
const client = require('prom-client');
new client.Summary({
  name: 'metric_name',
  help: 'metric_help',
  percentiles: [0.01, 0.1, 0.9, 0.99]
});
```

To enable the sliding window functionality for summaries you need to add
`maxAgeSeconds` and `ageBuckets` to the config like this:

```js
const client = require('prom-client');
new client.Summary({
  name: 'metric_name',
  help: 'metric_help',
  maxAgeSeconds: 600,
  ageBuckets: 5
});
```

The `maxAgeSeconds` will tell how old an bucket can be before it is reset and
`ageBuckets` configures how many buckets we will have in our sliding window for
the summary.

Usage example

```js
const client = require('prom-client');
const summary = new client.Summary({
  name: 'metric_name',
  help: 'metric_help'
});
summary.observe(10);
```

Utility to observe request durations

```js
const end = summary.startTimer();
xhrRequest(function(err, res) {
  end(); // Observes the value to xhrRequests duration in seconds
});
```

### Labels

All metrics can take a `labelNames` property in the configuration object. All
label names that the metric support needs to be declared here. There are two
ways to add values to the labels:

```js
const client = require('prom-client');
const gauge = new client.Gauge({
  name: 'metric_name',
  help: 'metric_help',
  labelNames: ['method', 'statusCode']
});

// 1st version: Set value to 100 with "method" set to "GET" and "statusCode" to "200"
gauge.set({ method: 'GET', statusCode: '200' }, 100);
// 2nd version: Same effect as above
gauge.labels('GET', '200').set(100);
```

It is also possible to use timers with labels, both before and after the timer
is created:

```js
const end = startTimer({ method: 'GET' }); // Set method to GET, we don't know statusCode yet
xhrRequest(function(err, res) {
  if (err) {
    end({ statusCode: '500' }); // Sets value to xhrRequest duration in seconds with statusCode 500
  } else {
    end({ statusCode: '200' }); // Sets value to xhrRequest duration in seconds with statusCode 200
  }
});
```

#### Default Labels (segmented by registry)

Static labels may be applied to every metric emitted by a registry:

```js
const client = require('prom-client');
const defaultLabels = { serviceName: 'api-v1' };
client.register.setDefaultLabels(defaultLabels);
```

This will output metrics in the following way:

```
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes{serviceName="api-v1"} 33853440 1498510040309
```

Default labels will be overridden if there is a name conflict.

`register.clear()` will clear default labels.

### Multiple registries

By default, metrics are automatically registered to the global registry (located
at `require('prom-client').register`). You can prevent this by specifying
`registers: []` in the metric constructor configuration.

Using non-global registries requires creating a Registry instance and passing it
inside `registers` in the metric configuration object. Alternatively you can
pass an empty `registers` array and register it manually.

Registry has a `merge` function that enables you to expose multiple registries
on the same endpoint. If the same metric name exists in both registries, an
error will be thrown.

```js
const client = require('prom-client');
const registry = new client.Registry();
const counter = new client.Counter({
  name: 'metric_name',
  help: 'metric_help',
  registers: [registry] // specify a non-default registry
});
const histogram = new client.Histogram({
  name: 'metric_name',
  help: 'metric_help',
  registers: [] // don't automatically register this metric
});
registry.registerMetric(histogram); // register metric manually
counter.inc();

const mergedRegistries = client.Registry.merge([registry, client.register]);
```

If you want to use multiple or non-default registries with the Node.js `cluster`
module, you will need to set the registry/registries to aggregate from:

```js
const AggregatorRegistry = client.AggregatorRegistry;
AggregatorRegistry.setRegistries(registry);
// or for multiple registries:
AggregatorRegistry.setRegistries([registry1, registry2]);
```

### Register

You can get all metrics by running `register.metrics()`, which will output a
string for prometheus to consume.

#### Getting a single metric for Prometheus displaying

If you need to output a single metric for Prometheus, you can use
`register.getSingleMetricAsString(*name of metric*)`, it will output a string
for Prometheus to consume.

#### Getting a single metric

If you need to get a reference to a previously registered metric, you can use
`register.getSingleMetric(*name of metric*)`.

#### Removing metrics

You can remove all metrics by calling `register.clear()`. You can also remove a
single metric by calling `register.removeSingleMetric(*name of metric*)`.

#### Resetting metrics

If you need to reset all metrics, you can use `register.resetMetrics()`. The
metrics will remain present in the register and can be used without the need to
instantiate them again, like you would need to do after `register.clear()`.

#### Cluster metrics

You can get aggregated metrics for all workers in a node.js cluster with
`register.clusterMetrics()`. This method both returns a promise and accepts a
callback, both of which resolve with a metrics string suitable for Prometheus to
consume.

```js
register
  .clusterMetrics()
  .then(metrics => {
    /* ... */
  })
  .catch(err => {
    /* ... */
  });

// - or -

register.clusterMetrics((err, metrics) => {
  // ...
});
```

### Pushgateway

It is possible to push metrics via a
[Pushgateway](https://github.com/prometheus/pushgateway).

```js
const client = require('prom-client');
let gateway = new client.Pushgateway('http://127.0.0.1:9091');

gateway.pushAdd({ jobName: 'test' }, function(err, resp, body) {}); //Add metric and overwrite old ones
gateway.push({ jobName: 'test' }, function(err, resp, body) {}); //Overwrite all metrics (use PUT)
gateway.delete({ jobName: 'test' }, function(err, resp, body) {}); //Delete all metrics for jobName

//All gateway requests can have groupings on it
gateway.pushAdd({ jobName: 'test', groupings: { key: 'value' } }, function(
  err,
  resp,
  body
) {});

//It's possible to extend the Pushgateway with request options from nodes core http/https library
gateway = new client.Pushgateway('http://127.0.0.1:9091', { timeout: 5000 }); //Set the request timeout to 5000ms
```

### Utilities

For convenience, there are 2 bucket generator functions - linear and
exponential.

```js
const client = require('prom-client');
new client.Histogram({
  name: 'metric_name',
  help: 'metric_help',
  buckets: client.linearBuckets(0, 10, 20) //Create 20 buckets, starting on 0 and a width of 10
});

new client.Histogram({
  name: 'metric_name',
  help: 'metric_help',
  buckets: client.exponentialBuckets(1, 2, 5) //Create 5 buckets, starting on 1 and with a factor of 2
});
```

The content-type prometheus expects is also exported as a constant, both on the
`register` and from the main file of this project, called `contentType`.

### Garbage Collection

To avoid native dependencies in this module, GC statistics for bytes reclaimed
in each GC sweep are kept in a separate module:
https://github.com/SimenB/node-prometheus-gc-stats.


CHANGES.md
# 2020-02-20, Version 12.0.0

- feature: remove all timestamp support (#333) (Sam Roberts)

- fix: collect metrics on scrape, not timeout (#329) (Sam Roberts)

- feat: implement advanced event loop monitoring (#278) (Yuriy Vasiyarov)

- fix: remove deprecated non-config object support (#330) (Sam Roberts)

- feat: implement GC metrics collection without native(C++) modules (#274) (Yuriy Vasiyarov)

- chore: add GH Actions CI (#323) (Simen Bekkhus)

- chore: update outdated dependencies (#322) (Simen Bekkhus)

- chore: spelling corrections in README (#325) (0xflotus)

- feat: Return the final duration value from startTimer callback (#282) (ricardoe)

- chore: add test for process_start_time_seconds (Sam Roberts)

- chore: correct var name in processStartTime (Sam Roberts)

- chore: remove ignored package-lock.json (#320) (Sam Roberts)

- chore: drop support for Node.js 8.x (appveyor) (#321) (Sam Roberts)

- chore: drop support for Node.js 8.x (#315) (Sam Roberts)

- fix: process_max_fds is process limit, not OS (#314) (Sam Roberts)

- chore: reindent package.json (Sam Roberts)

- Fix performance: avoid empty loops (Aleksei Androsov)

- fix: avoid mutation bug in registry.getMetricsAsJSON() (Aleksei Androsov)

- Cluster metrics aggregation ignores default labels (#170) (dependabot[bot])

- chore: fix CI/drop support for Node.js 6 (Zach Bjornson)

- Lazy-load cluster module to fix Passenger (Laurent Goudet)

- Bump mixin-deep from 1.3.1 to 1.3.2 (dependabot[bot])

- fix: TypeScript declaration - strict label values (#299) (Kobi)

- Messed up version number in Changelog (Simon Nyberg)

# 2019-06-27, Version 11.5.3

- Changelog 11.5.2 (Simon Nyberg)

- Compress t-digest to prevent memory leakage (#234) (Arne Schubert)

# 2019-06-20, Version 11.5.2

- fix: avoid mutation bug in registry (#273) (Jake Verbaten)

# 2019-06-13, Version 11.5.1

- fix: guard against missing constructor (Yuriy Vasiyarov)

# 2019-06-05, Version 11.5.0

- chore: prepare for publish (Simen Bekkhus)

- feat: add option to disable timestamps on defaultMetrics (#255) (Axel Dolce)

- Add validateMetricName() to the public interface (#246) (Paul Melnikow)

- chore: add recommended prettier eslint config (Simen Bekkhus)

- chore: bump dependencies (#269) (Simen Bekkhus)

- fix(types): incorrect return type of collectDefaultMetrics(), should be Timeout in Node (#264) (Nazar Mokrynskyi)

# 2019-06-04, Version 11.4.0

- chore: prepare for release (Simen Bekkhus)

- Split process request and process handles by type (#260) (Yuriy Vasiyarov)

# 2019-04-02, Version 11.3.0

- Changelog for 11.3.0 (Simon Nyberg)

- Fix rare ERR_IPC_CHANNEL_CLOSED bug in cluster mode (#244) (Zach Bjornson)

- Add remove() method for handling dynamic labels (#242) (Russell Centanni)

- Performance improvement (#235) (Denis Fäcke)

- removing the reset functionality from the readme from every single metric type (#243) (Krisztina Hirth)

- chore: test on node 8, not 9 [skip ci] (Simen Bekkhus)

- chore: fix appveyor build (#237) (Simen Bekkhus)

- chore: bump deps (Simen Bekkhus)

# 2018-12-21, Version 11.2.1

- Updated changelog (Simon Nyberg)

- update types for SummaryConfiguration (#231) (Gabriel Castro)

# 2018-11-13, Version 11.2.0

- Updated changelog for release (Simon Nyberg)

- Sliding windows for summaries (#229) (Oscar Tholander)

- Matt jarrett fix child dependency with vulnerability (#230) (Matt Jarrett)

- Create benchmark suite (#222) (Nowell Strite)

# 2018-09-22, Version 11.1.3

- prepare for 11.1.3 (Simen Bekkhus)

- Mutate labels to avoid excessive object cloning. (#220) (Nowell Strite)

- chore: add eslint-plugin-node (#221) (Simen Bekkhus)

# 2018-09-19, Version 11.1.2

- prepare for 11.1.2 (Simen Bekkhus)

- Fix histogram scrape performance (#219) (Nowell Strite)

- chore: upgrade prettier (#218) (Simen Bekkhus)

- Remove Sinon dev dependency (#217) (François Voron)

- Allow setting Gauge values to NaN. (#202) (Tim Joseph Dumol)

- adding prefix to DefaultMetricsCollectorConfiguration, related to the release v11.1.0. (#208) (Oscar Serna)

- Add getMetricsAsArray to index.d.ts (#211) (CubeKap)

- Updated README.md (#204) (Joseph Dalrymple)

# 2018-06-29, Version 11.1.1

- Prepare for 11.1.1 (Simon Nyberg)

- Fix #200 (#201) (François Voron)

# 2018-06-29, Version 11.1.0

- Prepare for 11.1.0 (Simon Nyberg)

- Add optional prefix to default metrics (#198) (Dan Forys)

- Fix aggregator method being lost when aggregating several registries (#193) (stupid cat)

- Validate labels on inc on counters (#190) (Ivan Vasiunyk)

- Fix `startTimer` utility to not mutate passed `startLabels` object (#188) (kdk)

- upgrade dependencies (#175) (Simen Bekkhus)

# 2018-03-10, Version 11.0.0

- Dropped node 4 support (#174) (Simon Nyberg)

- Use base units (seconds) for current time in gauge (#171) (Thom Wright)

# 2018-02-28, Version 10.2.3

- Added new version to changelog (Simon Nyberg)

- Cluster metrics aggregation ignores default labels (#170) (Aleksei Androsov)

- Use prettier on markdown and json (#164) (Simen Bekkhus)

# 2017-11-02, Version 10.2.2

- fixed invalid memory report on linux (#163) (Yurij Mikhalevich)

# 2017-10-27, Version 10.2.1

- Only resolve/reject `clusterMetrics` promise if no callback is provided (#162) (Zach Bjornson)

# 2017-10-16, Version 10.2.0

- Lazily initialize cluster event listeners (#159) (Zach Bjornson)

- fix for issue #156 (#158) (Jan Amoyo)

- reset Metrics (#157) (Erwin Poeze)

- Upgrade outdated dev dependencies (Simen Bekkhus)

# 2017-09-26, Version 10.1.1

- Bugfix - update TypeScript definitions (#154) (Tim Schoenheider)

- Bugfix - cannot get cluster info in node version < 6 (#153) (Simon Emms)

- Update links in changelog (Simen Bekkhus)

- Format changes in changelog (Simon Nyberg)

- Added new unrelased section in changelog (Simon Nyberg)

# 2017-09-04, Version 10.1.0

- Support aggregating metrics from a cluster (#147) (Zach Bjornson)

# 2017-08-22, Version 10.0.4


lib/summary.js
/**
 * Summary
 */
'use strict';

const util = require('util');
const type = 'summary';
const {
	getPropertiesFromObj,
	getLabels,
	hashObject,
	removeLabels
} = require('./util');
const { validateLabel } = require('./validation');
const { Metric } = require('./metric');
const timeWindowQuantiles = require('./timeWindowQuantiles');

const DEFAULT_COMPRESS_COUNT = 1000; // every 1000 measurements

class Summary extends Metric {
	constructor(config) {
		super(config, {
			percentiles: [0.01, 0.05, 0.5, 0.9, 0.95, 0.99, 0.999],
			compressCount: DEFAULT_COMPRESS_COUNT,
			hashMap: {}
		});

		for (const label of this.labelNames) {
			if (label === 'quantile')
				throw new Error('quantile is a reserved label keyword');
		}

		if (this.labelNames.length === 0) {
			this.hashMap = {
				[hashObject({})]: {
					labels: {},
					td: new timeWindowQuantiles(this.maxAgeSeconds, this.ageBuckets),
					count: 0,
					sum: 0
				}
			};
		}
	}

	/**
	 * Observe a value
	 * @param {object} labels - Object with labels where key is the label key and value is label value. Can only be one level deep
	 * @param {Number} value - Value to observe
	 * @returns {void}
	 */
	observe(labels, value) {
		observe.call(this, labels === 0 ? 0 : labels || {})(value);
	}

	get() {
		const data = getPropertiesFromObj(this.hashMap);
		const values = [];
		data.forEach(s => {
			extractSummariesForExport(s, this.percentiles).forEach(v => {
				values.push(v);
			});
			values.push(getSumForExport(s, this));
			values.push(getCountForExport(s, this));
		});

		return {
			name: this.name,
			help: this.help,
			type,
			values,
			aggregator: this.aggregator
		};
	}

	reset() {
		const data = getPropertiesFromObj(this.hashMap);
		data.forEach(s => {
			s.td.reset();
			s.count = 0;
			s.sum = 0;
		});
	}

	/**
	 * Start a timer that could be used to logging durations
	 * @param {object} labels - Object with labels where key is the label key and value is label value. Can only be one level deep
	 * @returns {function} - Function to invoke when you want to stop the timer and observe the duration in seconds
	 * @example
	 * var end = summary.startTimer();
	 * makeExpensiveXHRRequest(function(err, res) {
	 *	end(); //Observe the duration of expensiveXHRRequest
	 * });
	 */
	startTimer(labels) {
		return startTimer.call(this, labels)();
	}

	labels() {
		const labels = getLabels(this.labelNames, arguments);
		return {
			observe: observe.call(this, labels),
			startTimer: startTimer.call(this, labels)
		};
	}

	remove() {
		const labels = getLabels(this.labelNames, arguments);
		removeLabels.call(this, this.hashMap, labels);
	}
}

function extractSummariesForExport(summaryOfLabels, percentiles) {
	summaryOfLabels.td.compress();

	return percentiles.map(percentile => {
		const percentileValue = summaryOfLabels.td.percentile(percentile);
		return {
			labels: Object.assign({ quantile: percentile }, summaryOfLabels.labels),
			value: percentileValue ? percentileValue : 0
		};
	});
}

function getCountForExport(value, summary) {
	return {
		metricName: `${summary.name}_count`,
		labels: value.labels,
		value: value.count
	};
}

function getSumForExport(value, summary) {
	return {
		metricName: `${summary.name}_sum`,
		labels: value.labels,
		value: value.sum
	};
}

function startTimer(startLabels) {
	return () => {
		const start = process.hrtime();
		return endLabels => {
			const delta = process.hrtime(start);
			this.observe(
				Object.assign({}, startLabels, endLabels),
				delta[0] + delta[1] / 1e9
			);
		};
	};
}

function observe(labels) {
	return value => {
		const labelValuePair = convertLabelsAndValues(labels, value);

		validateLabel(this.labelNames, this.labels);
		if (!Number.isFinite(labelValuePair.value)) {
			throw new TypeError(
				`Value is not a valid number: ${util.format(labelValuePair.value)}`
			);
		}

		const hash = hashObject(labelValuePair.labels);
		let summaryOfLabel = this.hashMap[hash];
		if (!summaryOfLabel) {
			summaryOfLabel = {
				labels: labelValuePair.labels,
				td: new timeWindowQuantiles(this.maxAgeSeconds, this.ageBuckets),
				count: 0,
				sum: 0
			};
		}

		summaryOfLabel.td.push(labelValuePair.value);
		summaryOfLabel.count++;
		if (summaryOfLabel.count % this.compressCount === 0) {
			summaryOfLabel.td.compress();
		}
		summaryOfLabel.sum += labelValuePair.value;
		this.hashMap[hash] = summaryOfLabel;
	};
}

function convertLabelsAndValues(labels, value) {
	if (value === undefined) {
		return {
			value: labels,
			labels: {}
		};
	}

	return {
		labels,
		value
	};
}

module.exports = Summary;
